File: companies/IMC/39.combination-sum.go
package leetcode

import "sort"

// CombinationSum39 backtracks over sorted candidates with reuse allowed. Passing
// the same start index permits reuse; sorting lets the loop stop once val exceeds
// the remaining target.
//
// Time: exponential in output size. Space: O(target/min(candidate)) recursion.
func CombinationSum39(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	ans := [][]int{}
	path := []int{}
	var dfs func(int, int)
	dfs = func(start, remain int) {
		if remain == 0 {
			ans = append(ans, append([]int(nil), path...))
			return
		}
		for i := start; i < len(candidates); i++ {
			v := candidates[i]
			if v > remain {
				break
			}
			path = append(path, v)
			dfs(i, remain-v)
			path = path[:len(path)-1]
		}
	}
	dfs(0, target)
	return ans
}

File: Texas Instruments/39.combination-sum.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        /*
        Approach:
        Sort candidates and backtrack from a start index. Reusing the same
        candidate is allowed, so the recursive call keeps i as the next start.
        Stop a loop early once the candidate exceeds the remaining target.

        Complexity: Exponential output-sensitive time, O(target/min candidate)
        recursion depth excluding output.
        */
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> path;
        function<void(int, int)> dfs = [&](int start, int remain) {
            if (remain == 0) {
                ans.push_back(path);
                return;
            }
            for (int i = start; i < (int)candidates.size() && candidates[i] <= remain; ++i) {
                path.push_back(candidates[i]);
                dfs(i, remain - candidates[i]);
                path.pop_back();
            }
        };
        dfs(0, target);
        return ans;
    }
};
